src/LetsEncrypt.Azure/AzureKeyVaultCertificateRepository.cs

```diff
@@ -0,0 +1,143 @@
+        {
+            var certs = new List<X509Certificate2>();
+
+            foreach (var domain in _encryptOptions.Value.DomainNames)
+            {
+                var cert = await GetCertificateAsync(domain, cancellationToken);
+
+                if (cert != null)
+                {
+                    certs.Add(cert);
+                }
+            }
+
+            return certs;
+        }
+
+        private async Task<X509Certificate2?> GetCertificateAsync(string domain, CancellationToken token)
+        {
+            _logger.LogInformation("Searching for certificate in KeyVault for {Domain}", domain);
+
+            try
+            {
+                var normalizedName = NormalizeHostName(domain);
+                var certificate = await _client.GetCertificateAsync(normalizedName, token);
+
+                return new X509Certificate2(certificate.Value.Cer);
+            }
+            catch (RequestFailedException ex) when (ex.Status == 404)
+            {
+                _logger.LogWarning("Could not find certificate for {Domain} in Azure KeyVault", domain);
+            }
+            catch (CredentialUnavailableException ex)
+            {
+                _logger.LogError(ex, "Could not retrieve credentials for Azure Key Vault");
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Unexpected error attempting to retrieve certificate for {Domain} from Azure KeyVault. Verify settings and try again.", domain);
+            }
+
+            return null;
+        }
+
+        public async Task SaveAsync(X509Certificate2 certificate, CancellationToken cancellationToken)
+        {
+            var domainName = certificate.GetNameInfo(X509NameType.DnsName, false);
+
+            _logger.LogInformation("Saving certificate for {Domain} in Azure KeyVault.", domainName);
+
+            if (!(await ShouldImportVersionAsync(domainName, certificate, cancellationToken)))
+            {
+                _logger.LogInformation("Certificate for {Domain} is already up-to-date in Azure KeyVault. Skipping importing.", domainName);
+                return;
+            }
+
+            var exported = certificate.Export(X509ContentType.Pfx);
+
+            var options = new ImportCertificateOptions(NormalizeHostName(domainName), exported);
+
+            try
+            {
+                await _client.ImportCertificateAsync(options, cancellationToken);
+
+                _logger.LogInformation("Imported certificate into Azure KeyVault for {Domain}", domainName);
+            }
+            catch (RequestFailedException ex)
+            {
+                _logger.LogWarning(ex, "Failed to save {Domain} certificate to Azure KeyVault", domainName);
+            }
+        }
+
+        private async ValueTask<bool> ShouldImportVersionAsync(string domainName, X509Certificate2 certificate, CancellationToken token)
```

Comment:
Maybe I missed something, but this is unlikely, right? Do we need this check? When would we be pulling a cert from Let's Encrypt that is already provided by Azure Key Vault?

Reply:
The saving method is called even when it's pulled from a source. So, if it is pulled from keyvault, this check is to prevent it from being imported again when it's already there. So, maybe a better fix would be to only call the save method when new certificate is created from LetsEncrypt, but I put this in there as a stop gap.